### PR TITLE
Improved Save Times

### PR DESCRIPTION
**Pull request by Sarkath**
- What is the purpose of the pull request?

This makes some optimizations to the `GameCharacter.getChildren()` and `GameCharacter.getNonCommonNodes()` to decrease save times by around 33%. This change is most noticeable on games with a large number of offspring.

- Give a brief description of what you changed or added.

I replaced the Java streams in those functions with basic loops. Based on my benchmarks, loops are noticeably more performant in this situation than sequential streams (and this particular case doesn't seem to play well with parallel streams).

On my [stress test save](https://files.sarkath.net/liliths_throne/StressTest/BreedingSave2.7z) (containing the end result of a manor full of bunnies breeding with a broodmother pill active…twice), this brought the average save time from 8.740s to 5.894s (determined by saving the game five times, dropping the highest and lowest times, and averaging the three remaining values).

- Are any new graphical assets required?

No.

- Has this change been tested? If so, mention the version number that the test was based on.

This was tested against 0.3.20.

**Edit:** As of 2021-05-14, this has also been merged with and tested against 0.4.

- So we have a better idea of who you are, what is your Discord Handle?

I'm not on Discord. :eyes:

### CODE REVIEW

File: src/com/lilithsthrone/game/character/GameCharacter.java [hunk 1]
```diff
@@ -5336,20 +5336,30 @@ public String getCompanionSexRejectionReason(boolean companionIsSub) {
...
+		for(Litter litter : getLittersBirthed()) {
+			for(String id : litter.getOffspring()) {
+				try {
+					gc = Main.game.getNPCById(id);
+				} catch (Exception ex) {
+					continue;
+				}
+				result.add(gc);
+			}
+		}
```
Comment (CognitiveMist) [inline src/com/lilithsthrone/game/character/GameCharacter.java]: These could be reduced like so, eliminating the need for the temp var `gc`:
```diff
 		for(Litter litter : getLittersBirthed()) {
 			for(String id : litter.getOffspring()) {
 				try {
-					gc = Main.game.getNPCById(id);
+					result.add(Main.game.getNPCById(id));
 				} catch (Exception ex) {
 					continue;
 				}
-				result.add(gc);
 			}
 		}
```

Comment (Sarkath) [reply src/com/lilithsthrone/game/character/GameCharacter.java]: My bad! I was prototyping some more aggressive changes (ones that would impact other parts of the engine with minimal improvement to save times) that would have required a temporary variable and forgot to change it back. Pushing an update now.

**Comment by bicobus** [comment]
> This change is most noticeable on games with a large number of offspring.

Could you please expend on what you mean by a *large* number? How large does it needs to be before becoming an issue?

**Comment by Sarkath** [comment]
My stress test save has a a few thousand NPCs (close to 4000, if memory serves), but I noticed times starting to creep over a second or two with as few as 500-700 active NPCs (which I managed to hit more or less by accident on my first playthrough…whoopsie).

This mostly just affects runs with a bunch of offspring due to Domination having relatively limited space, though I imagine that it will become a more prominent issue as the game world expands and is forced to track a larger number of NPCs.

**Comment by Maxis010** [comment]
Tested on version 0.4 character with 6835 NPCs
Save file size 106Mb (Approx 2 million lines)

Innoxia dev branch 37 seconds average from 5 tests
Sarkath feature/save-perf-improvement branch 18 seconds average from 5 tests

No issues loading the save afterwards and no missing data when reviewing the save files

**Comment by Innoxia** [comment]
Thank  you for this! If you decide to join the LT discord server, throw me a PM (or ping me) and I'll add you to the developers channel.

**Pull request merged**